agent/config/file_watcher.go

```diff
@@ -100,25 +107,51 @@ func (w *FileWatcher) add(filename string) error {
 	if err != nil {
 		return err
 	}
+	w.configFilesLock.Lock()
+	defer w.configFilesLock.Unlock()
 	w.configFiles[filename] = &watchedFile{modTime: modTime}
 	return nil
 }
 
-func isSymLink(filename string) bool {
-	fi, err := os.Lstat(filename)
-	if err != nil {
-		return false
+// Remove a file from the file watcher
+// Remove will lock the file watcher during the remove
+func (w *fileWatcher) Remove(filename string) {
+	w.configFilesLock.Lock()
+	defer w.configFilesLock.Unlock()
+	delete(w.configFiles, filename)
+}
+
+// Replace a file in the file watcher
+// Replace will lock the file watcher during the replace
+func (w *fileWatcher) Replace(oldFile, newFile string) error {
+	if oldFile == newFile {
+		return nil
 	}
-	if fi.Mode()&os.ModeSymlink != 0 {
-		return true
+	newFile = filepath.Clean(newFile)
+	w.logger.Trace("adding file", "file", newFile)
+	if err := w.watcher.Add(newFile); err != nil {
+		return err
+	}
+	modTime, err := w.getFileModifiedTime(newFile)
+	if err != nil {
+		return err
 	}
-	return false
+	w.configFilesLock.Lock()
+	defer w.configFilesLock.Unlock()
+	delete(w.configFiles, oldFile)
+	w.configFiles[newFile] = &watchedFile{modTime: modTime}
```

Comment:
_(purely optional; codegolf)_
You could move just the interaction of the lock and the map into some helpers to keep them tighter, like:
```
func (w *fileWatcher) replaceFile(oldFile, newFile string, modTime time.Time) {
	w.configFilesLock.Lock()
	defer w.configFilesLock.Unlock()
	delete(w.configFiles, oldFile)
	w.configFiles[newFile] = &watchedFile{modTime: modTime}
}

func (w *fileWatcher) addFile(newFile string, modTime time.Time) {
	w.configFilesLock.Lock()
	defer w.configFilesLock.Unlock()
	w.configFiles[filename] = &watchedFile{modTime: modTime}
}

func (w *fileWatcher) removeFile(oldFile string) {
	w.configFilesLock.Lock()
	defer w.configFilesLock.Unlock()
	delete(w.configFiles, oldFile)
}
```